flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaShuffleFetcher.java

```diff
@@ -0,0 +1,344 @@
+
+		public boolean isRecord() {
+			return getClass() == KafkaShuffleRecord.class;
+		}
+
+		boolean isWatermark() {
+			return getClass() == KafkaShuffleWatermark.class;
+		}
+
+		KafkaShuffleRecord<T> asRecord() {
+			return (KafkaShuffleRecord<T>) this;
+		}
+
+		KafkaShuffleWatermark asWatermark() {
+			return (KafkaShuffleWatermark) this;
+		}
+	}
+
+	private static class KafkaShuffleWatermark<T> extends KafkaShuffleElement<T> {
+		final int subtask;
+		final long watermark;
+
+		KafkaShuffleWatermark(int subtask, long watermark) {
+			this.subtask = subtask;
+			this.watermark = watermark;
+		}
+	}
+
+	private static class KafkaShuffleRecord<T> extends KafkaShuffleElement<T> {
+		final T value;
+		final Long timestamp;
+
+		KafkaShuffleRecord(T value) {
+			this.value = value;
+			this.timestamp = null;
+		}
+
+		KafkaShuffleRecord(long timestamp, T value) {
+			this.value = value;
+			this.timestamp = timestamp;
+		}
+	}
+
+	private static class KafkaShuffleElementDeserializer<T> implements Serializable {
+		private transient DataInputDeserializer dis;
+
+		KafkaShuffleElementDeserializer() {
+			this.dis = new DataInputDeserializer();
+		}
+
+		KafkaShuffleElement<T> deserialize(TypeSerializer<T> serializer, ConsumerRecord<byte[], byte[]> record)
+			throws Exception {
+			byte[] value = record.value();
+			dis.setBuffer(value);
+			int tag = IntSerializer.INSTANCE.deserialize(dis);
+
+			if (tag == TAG_REC_WITHOUT_TIMESTAMP) {
+				return new KafkaShuffleRecord<>(serializer.deserialize(dis));
+			} else if (tag == TAG_REC_WITH_TIMESTAMP) {
+				return new KafkaShuffleRecord<>(LongSerializer.INSTANCE.deserialize(dis), serializer.deserialize(dis));
+			} else if (tag == TAG_WATERMARK) {
+				return new KafkaShuffleWatermark<>(
+					IntSerializer.INSTANCE.deserialize(dis), LongSerializer.INSTANCE.deserialize(dis));
+			}
+
+			throw new UnsupportedOperationException("Unsupported tag format");
+		}
+	}
+
+	/**
+	 * WatermarkHandler to generate watermarks.
+	 */
+	private static class WatermarkHandler {
+		private final int producerParallelism;
+		private final Map<Integer, Long> subtaskWatermark;
+
+		private long currentMinWatermark = Long.MIN_VALUE;
+
+		WatermarkHandler(int numberOfSubtask) {
+			this.producerParallelism = numberOfSubtask;
+			this.subtaskWatermark = new HashMap<>(numberOfSubtask);
+		}
+
+		public Optional<Watermark> checkAndGetNewWatermark(KafkaShuffleWatermark newWatermark) {
+			// watermarks is incremental for the same partition and PRODUCER subtask
+			Long currentSubTaskWatermark = subtaskWatermark.get(newWatermark.subtask);
+
+			Preconditions.checkState(
+				(currentSubTaskWatermark == null) || (currentSubTaskWatermark <= newWatermark.watermark),
+				"Watermark should always increase");
+
+			subtaskWatermark.put(newWatermark.subtask, newWatermark.watermark);
+
+			if (subtaskWatermark.values().size() < producerParallelism) {
+				return Optional.empty();
+			}
```

Comment:
What happens if one partition has ended and we receive no watermarks anymore? Are the watermarks of the other partitions still propagated properly? Almost feels like using `StatusWatermarkValve` would be handy.

Reply:
Yes, they are very similar at first glance. But my feeling is `StatusWatermarkValve` is an overkill since it also includes StreamStatus. And the pick up of watermark depends on the stream status.

- If one partition has ended and we receive no watermarks anymore, are the watermarks of the other partitions still propagated properly?

As long as all subtasks contain watermarks, the watermark progress properly. The only problem is if one of the subtask does not produce watermark. But that's a common question to find a min/max watermark amongst all channels. 

Data is propogates fine.

Reply:
I'd assume you need to propagate EndOfPartitionEvents downstream and adjust `producerParallelism` accordingly.